python/labbox_ephys/helpers/prepare_snippets_h5.py

```diff
@@ -101,3 +113,41 @@ def prepare_snippets_h5_from_extractors(
             
+def prepare_snippets_nwb_from_extractors(
+    recording: se.RecordingExtractor,
+    sorting: se.SortingExtractor,
+    nwb_file_path: str,
+    nwb_object_prefix: str,
+    start_frame,
+    end_frame,
+    max_neighborhood_size: int,
+    max_events_per_unit: Union[None, int]=None,
+    snippet_len=(50, 80),
+):
+    import pynwb
+    samplerate, unit_ids, unit_waveforms, channel_ids_by_unit, sorting_subsampled = _prepare_snippets_h5_from_extractors_helper(
+        recording=recording,
+        sorting=sorting,
+        start_frame=start_frame,
+        end_frame=end_frame,
+        max_neighborhood_size=max_neighborhood_size,
+        max_events_per_unit=max_events_per_unit,
+        snippet_len=snippet_len
+    )
+    # thank you Ryan Ly
+    with pynwb.NWBHDF5IO(path=nwb_file_path, mode='a') as io:
+        nwbf = io.read()
+        nwbf.add_scratch(name=f'{nwb_object_prefix}_unit_ids', data=np.array(unit_ids).astype(np.int32), notes='sorted waveform unit ids')
+        nwbf.add_scratch(name=f'{nwb_object_prefix}_sampling_frequency', data=np.array([samplerate]).astype(np.float64), notes='sorted waveform sampling frequency')
+        nwbf.add_scratch(name=f'{nwb_object_prefix}_channel_ids', data=np.array(recording.get_channel_ids()), notes='sorted waveform channel ids')
+        nwbf.add_scratch(name=f'{nwb_object_prefix}_num_frames', data=np.array([recording.get_num_frames()]).astype(np.int32), notes='sorted waveform number of frames')
+        channel_locations = recording.get_channel_locations()
+        nwbf.add_scratch(name=f'{nwb_object_prefix}_channel_locations', data=np.array(channel_locations), notes='sorted waveform channel locations')
+        for ii, unit_id in enumerate(unit_ids):
+            x = sorting.get_unit_spike_train(unit_id=unit_id)
+            nwbf.add_scratch(name=f'{nwb_object_prefix}_unit_{unit_id}_spike_trains', data=np.array(x).astype(np.float64), notes=f'sorted spike trains for unit {unit_id}')
+            nwbf.add_scratch(name=f'{nwb_object_prefix}_unit_{unit_id}_waveforms', data=unit_waveforms[ii].astype(np.float32), notes=f'sorted waveforms for unit {unit_id}')
+            nwbf.add_scratch(name=f'{nwb_object_prefix}_unit_{unit_id}_channel_ids', data=np.array(channel_ids_by_unit[int(unit_id)]).astype(int), notes=f'sorted channel ids for unit {unit_id}')
+            nwbf.add_scratch(name=f'{nwb_object_prefix}_unit_{unit_id}_sub_spike_train', data=np.array(sorting_subsampled.get_unit_spike_train(unit_id=unit_id)).astype(np.float64), notes=f'sorted subsampled spike train for unit {unit_id}')
+        io.write(nwbf)
+    # caveat: The above code working is dependent on a new release of PyNWB/HDMF that would allow the storage of scalars in the scratch space: NeurodataWithoutBorders/pynwb#1309 . This code works on dev versions of nwb_datajoint and labbox-ephys.
```

Comment:
```suggestion
    with pynwb.NWBHDF5IO(path=nwb_file_path, mode='a') as io:
        nwbf = io.read()
        if not is_noise_sorting:
            if nwbf.units is None:
                nwbf.units = pynwb.misc.Units(name='units')
            units_table = nwbf.units
        else:
            units_table = pynwb.misc.Units(
                name='NoiseUnits',
                description='noise units',
                electrode_table=nwbf.electrodes
            )
            nwbf.add_scratch(units_table)

        if 'sub_spike_times' not in units_table:
            units_table.add_column(name='sub_spike_times', description='subsampled spike times', index=True)
        if 'waveforms' not in units_table:
            units_table.add_column(name='waveforms', description='spike waveforms', index=2)
        if 'sort_interval' not in units_table:
            units_table.add_column(name='sort_interval', description='the interval used for spike sorting')
        # electrodes with id equal to those in channel_ids_by_unit must already exist in the file
        electrode_row_map = {id: ii for ii, id in enumerate(nwbf.electrodes.id.data)}
        for ii, unit_id in enumerate(unit_ids):
            electrode_inds = [electrode_row_map[a] for a in channel_ids_by_unit[int(unit_id)]]
            units_table.add_row(
                id=unit_id,
                spike_times=sorting.get_unit_spike_train(unit_id=unit_id).astype(np.float64),
                sub_spike_times=sorting_subsampled.get_unit_spike_train(unit_id=unit_id).astype(np.float64),
                waveforms=unit_waveforms[ii].astype(np.float32),
                electrodes=electrode_inds,
                sort_interval=sort_interval,
                obs_intervals=sort_interval_valid_times,
                waveform_mean=templates[ii],
            )
            # TODO copy the file to a new file and add waveforms to that file
        units_table.resolution = 1/samplerate
        io.write(nwbf)
        return units_table.object_id
```
This code uses the Units type in NWB to store the sorted units information rather than putting it in the scratch space as I had initially proposed.